File: problemList/slidingWindow/solution/LC1493.java
package problemList.slidingWindow.solution;

/**
题意: 至多包含一个0的最长子数组长度
 */
public class LC1493 {
    public int longestSubarray(int[] nums) {
        int n = nums.length, ret = 0, zeroCnt = 0, left = 0;
        for(int i = 0;i < n;i++){
            zeroCnt += (1 - nums[i]);
            while(zeroCnt > 1){
                zeroCnt -= (1 - nums[left]);
                left++;
            }
            ret = Math.max(ret, i - left + 1);
        }
        return ret - 1;
    }
}

File: SolutionsWithCorCpp/problem_list_0x3f/sliding_window/solution/LC1493.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

// --> 只包含一个0的最长子数组长度
int longestSubarray(vector<int> &nums) {
    int n = nums.size(), r = 0, ret = 0, cnt = 0;
    for(int i = 0;r < n;r++){
        if(nums[r] == 0) cnt++;
        while(cnt > 1 && i < n){
            if(nums[i] == 0) cnt--;
            i++;
        }
        if(i <= r) ret = max(ret, r - i + 1);
    }
    return min(ret - 1, n - 1);
}

int main(){
    vector<int> nums = {0,1,1,1,0,1,1,0,1};
    cout << longestSubarray(nums) << endl;
    return 0;
}